bindings/csharp/Engine.cs

```diff
@@ -0,0 +1,114 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+
+namespace Keystone
+{
+    public class Engine
+    {
+        private IntPtr engine = IntPtr.Zero;
+
+        public IntPtr Pointer()
+        {
+            return engine;
+        }
+
+        public bool IsReady()
+        {
+            return (engine != IntPtr.Zero);
+        }
+
+        public uint Version(ref int major, ref int minor)
+        {
+            return (uint) Import.ks_version(ref major, ref minor);
+        }
+
+        public bool IsArchSupported(Constants.ks_arch arch)
+        {
+            return Import.ks_arch_supported(arch);
+        }
+
+        public Constants.ks_err Open(Constants.ks_arch arch, Constants.ks_mode mode)
+        {
+            return Import.ks_open(arch, (int) mode, ref engine);
+        }
+
+        public Constants.ks_err Close()
+        {
+            return Import.ks_close(engine);
+        }
+
+        public Constants.ks_err Errno()
+        {
+            return Import.ks_errno(engine);
+        }
+
+        public string StrError(Constants.ks_err code)
+        {
+            return Import.ks_strerror(code);
+        }
+
+        public Constants.ks_err SetSymbolResolver(Constants.ks_sym_resolver resolver)
+        {
+            return Import.ks_option(engine, Constants.ks_opt_type.KS_OPT_SYM_RESOLVER, Marshal.GetFunctionPointerForDelegate(resolver));
```

Comment:
This will crash on garbage collection. The function `Marshal.GetFunctionPointerForDelegate` doesn't increase the reference count so if you pass a lambda to `SetSymbolResolver` and you call `GC.Collect()` the object will hold no reference and the delegate will be deleted. A solution is to store a private copy `Constants.ks_sym_resolver` inside this class to increase the reference count. This should be set to `null` on disposal of the `engine`.